tidb-cascades-planner.md

```diff
@@ -0,0 +1,399 @@
+
+Cascades Optimizer 为 Rule 的应用顺序做了很细致的设计，例如每个 Rule 都有 promise 和 condition 两个方法，其中 promise 用来表示 Rule 在当前搜索过程中的重要性，promise 值越高，则该规则越可能有用，当 promise 值小于等于 0 时，这个 Rule 就不会被执行；而 condition 直接通过返回一个布尔值决定一个 Rule 是否可以在当前过程中被应用。当一个 Rule 被成功应用之后，会计算下一步有可能会被应用的 Rule 的集合。
+
+Volcano/Cascades Optimzier 都使用了 Branch-And-Bound 的方法对搜索空间进行剪枝。由于两者都采用了自顶向下的搜索，在搜索的过程中可以为算子设置其 Cost Upper Bound，如果在向下搜索的过程中还没有搜索到叶子节点就超过了预设的 Cost Upper Bound，就可以对这个搜索分支预先进行剪枝。
+
+## TiDB Cascades Planner 的设计
+
+基于 Volcano/Cascades Optimizer 的原理，我们为 TiDB 重新设计了一个优化器：TiDB Cascades Planner。我们希望可以通过新的优化器来解决现行优化器的问题，并且也能够带来一些新的特性：
+
+* 优化规则易于实现，通过实现几个简单的接口来定义优化规则。
+* 优化规则易于扩展，我们不需要再考虑优化规则的执行顺序。
+* 优化规则可以反复应用，增大优化器的搜索空间。
+* 对于不一定更优的优化规则，可以通过 Cost 来选取结果。
+* 算子下推存储层更加灵活，方便未来扩展新的下推算子。
+* 使 TiDB 可以更容易地接入其他的存储或者计算引擎，例如 [TiFlash](https://pingcap.com/blog-cn/tidb-with-tiflash-extension/)。
+* 为 TiDB 的优化器能力分级，不同复杂程度的查询可以选用不同的优化等级。
+
+### 基本数据结构
+
+首先我们介绍 TiDB Cascades Planner 中的一些基本的数据结构，以下的部分概念与上文介绍的 Volcano/Cascades Optimizer 大体一致，只会有少许的不同。
+
+#### Group/GroupExpr
+
+GroupExpr 是对 `LogicalPlan` 的封装，与 `LogicalPlan` 不同的是，GroupExpr 的子节点不再是 `LogicalPlan`，而是 Group：
+
+```golang
+type GroupExpr struct {
+  ExprNode plannercore.LogicalPlan
+  Children []*Group
+  Group    *Group
+  ...
+}
+```
+
+Group 是一组逻辑等价的 GroupExpr 集合，换句话说，从逻辑上来看，通过一个 Group 中任何一个 GroupExpr 产生的算子树都是逻辑等价的。
+
+```golang
+type Group struct {
+  Equivalents *list.List 
+
+  ImplMap map[string]Implementation
+  Prop    *property.LogicalProperty 
+  EngineType EngineType
+}
+```
+
+为了加快搜索的过程，我们对 Group 做了一些更细粒度的优化，例如在将 GroupExpr 插入到 Equivalents 时，我们总是保证相同类型的 LogicalPlan 在链表中连续存储；同时为每种类型的首个 GroupExpr 提供一个 map 作为索引等。
+
+通过以上两个定义我们可以发现，Group 和 GroupExpr 相互递归地引用，最终形成一个 Memo 数据结构。
+
+#### Operand
+
+Operand 是 LogicalPlan 的类型符，用于描述 Pattern。
+
+```golang
+type Operand int
+const (
+  OperandAny Operand = iota
+  OperandJoin
+  OperandAggregation
+  OperandProjection
+  ...
+```
+
+#### Pattern
+
+Pattern 是一个树状的数据结构，用于表示逻辑算子树的局部的形状。需要注意的是 Pattern 只能用于匹配逻辑算子的类型（通过 Operand），但是不能够指定算子中具体的内容。
+
+```golang
+type Pattern struct {
+  Operand
+  Children []*Pattern
+}
+```
+
+#### Transformation
+
+Transformation 是一个接口类型，用来定义一个逻辑变换规则。
+
+*   `GetPattern()` 方法获取这个变换规则所需要匹配的一个 Pattern。
+
+*   由于 Pattern 只能描述算子的类型，不能描述 LogicalPlan 内部的内容约束，因此通过 `Match()` 方法可以判断更细节的匹配条件。例如 Pattern 只能描述我们想要一个 Join 类型的算子，但是却没法描述这个 Join 应该是 InnerJoin 或者是 LeftOuterJoin，这类条件就需要在 `Match()` 中进行判断。
+
+*   `OnTransform()` 方法中定义了变换规则的具体内容，返回的内容分别是新的 GroupExpr，是否删除旧的 `GroupExpr`，是否删除旧的 Group 中所有的 `GroupExpr`。
+
+```golang
+// Transformation defines the interface for the transformation rules.
+type Transformation interface {
+  GetPattern() *memo.Pattern
+ 
+  Match(expr *memo.ExprIter) bool
+  
+  OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error)
+}
+```
+
+
+下面我们以一个变换规则：`PushSelDownAggregation` 为例，具体介绍上面三个方法的使用方式。
```

Comment:
Add a link https://github.com/pingcap/tidb/blob/6a5955750014f239a41362059ced6d8ab420f7b4/planner/cascades/transformation_rules.go#L394 for `PushSelDownAggregation`.